### PR TITLE
Move gradle build scripts to 100% Kotlin

### LINKED ISSUE #17: Convert gradle files to kotlin

### PR DESCRIPTION
**Pull request by 0legg**
Resolve #17 (with few questionable parts though). Downsides — Android Studio tends to go mad from any build.gradle.kts. CLI is fine, IntelliJ IDEA works well until it fails to use state-of-art Android Gradle plugin.

### CODE REVIEW

File: buildSrc/src/main/java/dependencies.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/* ktlint-disable no-multi-spaces max-line-length */
+object deps {
```
Comment (hzsweers) [inline buildSrc/src/main/java/dependencies.kt]: should these all be `const` too?

File: retrograde-util/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    implementation(deps.libs.koptionalRxJava2)
+    implementation(deps.libs.kotlinStdlib)
+    implementation(deps.libs.okHttp3)
+    implementation(deps.libs.roomRuntime)
+    implementation(deps.libs.rxJava2)
+    implementation(deps.libs.supportAppCompatV7)
+    implementation(deps.libs.supportRecyclerViewV7)
+}
+
+android { }
```
Comment (hzsweers) [inline retrograde-util/build.gradle.kts]: Is this necessary?

File: build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+configurations {
+    maybeCreate("ktlint")
+}
+
+dependencies {
+    "ktlint"("com.github.shyiko:ktlint:0.14.0")
+}
+
+tasks {
+    "clean"(Delete::class) {
```
Comment (hzsweers) [inline build.gradle.kts]: dafuq kind of syntax is this. I hope this extension method is just inside the tasks scope...

Comment (0legg) [reply build.gradle.kts]: Looks weird, borrowed it from few examples in gradle repo. As soon as there is no well-established gradle-kotlin-script guidelines, I'd stick with this approach. Looks somewhat groovyish.

Yes, it's inside tasks scope — https://github.com/gradle/kotlin-dsl/blob/master/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt

File: buildSrc/src/main/java/dependencies.kt (hunk 1 above)
Comment (0legg) [reply buildSrc/src/main/java/dependencies.kt]: Good catch

File: retrograde-util/build.gradle.kts (hunk 2 above)
Comment (0legg) [reply retrograde-util/build.gradle.kts]: I kept original files structure, it can be easily removed

**Comment by codebutler** [comment]
This is awesome, thanks a lot! I'm concerned about merging if Android Studio is going to go crazy. Are those issues known and being tracked somewhere upstream?

File: build.gradle.kts (hunk 3 above)
Comment (codebutler) [reply build.gradle.kts]: Is this really the recommended syntax?

**Comment by 0legg** [comment]
Speaking of issues — probably, I was over-dramatizing on Android Studio reaction to Gradle Kotlin DSL. The only downside I see so far is that if you have some error in your code, then half of your Gradle file will go red. It's pretty useful, at least it's good enough for 0.14.x version.

File: build.gradle.kts (hunk 3 above)
Comment (0legg) [reply build.gradle.kts]: There is no well-established syntax yet, but latest samples are intensively using this syntax: https://github.com/gradle/kotlin-dsl/blob/a5de7d1b485ae30755d4ac09e54063cd9e9678e3/samples/task-dependencies/build.gradle.kts

**Comment by tyvsmith** [comment]
Thanks @0legg, great work. @codebutler I'd recommend merging with the AS issue, as it's primarily around handling of non-compilable code.

**Comment by codebutler** [comment]
Thanks everyone!!

**Comment by hzsweers** [comment]
FYI, seems to be broken on AGP 3.1.x - https://github.com/gradle/kotlin-dsl/issues/708

**Pull request merged**